### PR TITLE
Optimize Permute Kernel in DeepEP

### PR DESCRIPTION
**Pull request by xutizhou**
## Motivation

The current performance of DeepEP is suboptimal due to the low efficiency of PyTorch's native permute function, which is used for formatting data before and after DeepEP communication. To address this limitation, we have implemented high-efficiency Triton kernels that significantly improve overall performance.

```
Co-authored-by: @zhou9402
```
#### Single Node
Command
```
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code --tp 8 --dp 8 --host 0.0.0.0 --port 30000 --enable-dp-attention --enable-deepep-moe --max-running-requests 128 --disable-radix-cache --mem-fraction-static 0.9 --stream-output --disable-cuda-graph

python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompt 512 --random-input 1000 --random-output 1000 --random-range-ratio 1 --host 127.0.0.1 --port 30000 --max-concurrency 128
```
|Version | Concurrency | Input | Output | Num Requests | Input Throughput(tok/s) | Output Throughput (tok/s) | Total Throughput (tok/s) |
|-----|-----|-----|-----|-----|-----|-----|-----|
| DeepEP(original) | 127.97 | 1000 | 1000 | 512 | 436.69 | 436.69 | 873.38 |
| DeepEP(current) | 127.97 | 1000 | 1000 | 512 | 581.94 | 581.94 | 1163.87  |

#### Multi Node
Command
```
# node 0
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code \
  --tp 16 --dp 16  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 0 \
  --enable-dp-attention --enable-deepep-moe \
  --disable-cuda-graph
# node 1
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code \
  --tp 16 --dp 16  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 1 \
  --enable-dp-attention --enable-deepep-moe \
  --disable-cuda-graph
```
|Version | Concurrency | Input | Output | Num Requests | Input Throughput(tok/s) | Output Throughput (tok/s) | Total Throughput (tok/s) |
|-----|-----|-----|-----|-----|-----|-----|-----|
| DeepEP(current) | 255.93 | 1000 | 1000 | 512 |956.36 | 956.36 | 1912.71 |
| DeepEP(current) | 511.31 | 1000 | 1000 | 1024 |1711.54 | 1711.54 | 3423.09 |
| DeepEP(current) | 1023.17 | 1000 | 1000 | 2048 |2974.21 | 2974.21 | 5948.42 |
| DeepEP(current) | 2046.18 | 1000 | 1000 | 4096 |3929.73 | 3929.73 | 7859.46 |
| EPMoe | 255.55 | 1000 | 1000 | 512 |868.55 | 868.55 | 1737.10 |
| EPMoe | 511.85 | 1000 | 1000 | 1024 |1694.59 | 1694.59 | 3389.18 |
| EPMoe | 1022.27 | 1000 | 1000 | 2048 |2735.53 | 2735.53 | 5471.06 |
| EPMoe | 2045.90 | 1000 | 1000 | 4096 |3489.57 | 3489.57 | 6979.15 |
## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 1]
```diff
@@ -17,6 +17,116 @@
...
+    dst_id = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = dst_id < num_toks
+    src_id = tl.load(reorder_ids + dst_id, mask=mask)
+    num_invalid = tl.load(num_minus_one)
+    tl.store(src2dst + src_id, dst_id - num_invalid, mask=mask)
+
+
+def deepep_run_moe_deep_preprocess(topk_ids: torch.Tensor, num_experts: int):
+    reorder_topk_ids, reorder_ids = torch.sort(topk_ids.view(-1), stable=True)
+    seg_indptr = torch.zeros(num_experts + 1, device=topk_ids.device, dtype=torch.int64)
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: It can be init using `torch.empty`

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 2]
```diff
@@ -17,6 +17,116 @@
...
+    torch.searchsorted(reorder_topk_ids, expert_ids, out=seg_indptr)
+    num_minus_one = seg_indptr[0]
+    seg_indptr = seg_indptr - num_minus_one
+
+    BLOCK_SIZE = 512
+    grid = (triton.cdiv(topk_ids.numel(), BLOCK_SIZE),)
+    deepep_compute_src2dst_triton_kernel[grid](
+        reorder_ids, src2dst, topk_ids.numel(), num_minus_one, BLOCK_SIZE
+    )
+    # src2dst -= num_minus_one
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: debugging code?

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 3]
```diff
@@ -17,6 +17,116 @@
 logger = logging.getLogger(__name__)
 
 
+@triton.jit
+def compute_src2dst_triton_kernel(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: `compute_src2dst_triton_kernel` and `deepep_compute_src2dst_triton_kernel` are defined twice.

Comment (ch-wan) [reply python/sglang/srt/layers/moe/ep_moe/kernels.py:21]: It is equivalent to this pure PyTorch code:

```python
src2dst[reorder_ids] = torch.arange(num_toks, device=reorder_ids.device)
```

Why developing a triton kernel is necessary? Is it faster?

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 4]
```diff
@@ -17,6 +17,116 @@
...
+):
+    pid = tl.program_id(axis=0)
+    dst_id = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = dst_id < num_toks
+    src_id = tl.load(reorder_ids + dst_id, mask=mask)
+    tl.store(src2dst + src_id, dst_id, mask=mask)
+
+
+@triton.jit
+def deepep_compute_src2dst_triton_kernel(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: Why developing a triton kernel is necessary? Is it faster?

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 5]
```diff
@@ -17,6 +17,116 @@
...
+    deepep_compute_src2dst_triton_kernel[grid](
+        reorder_ids, src2dst, topk_ids.numel(), num_minus_one, BLOCK_SIZE
+    )
+    # src2dst -= num_minus_one
+    reorder_topk_ids = reorder_topk_ids[num_minus_one:]
+    return reorder_topk_ids, src2dst, seg_indptr
+
+
+@triton.jit
+def deepep_permute_triton_kernel(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py:21]: It is defined twice.

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 6]
```diff
@@ -17,6 +17,116 @@
...
+
+        for idx in range(topk):
+            dst_idx = tl.load(src2dst_ptr + idx)
+            if dst_idx >= 0:
+                dst_ptr = gateup_input_ptr + dst_idx * hidden_size
+                tl.store(dst_ptr + offset, out_data, mask=mask)
+
+
+@triton.jit
+def deepep_post_reorder_triton_kernel(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: It is defined twice.

**Review by ch-wan — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/moe/ep_moe/token_dispatcher.py [hunk 7]
```diff
@@ -427,10 +347,29 @@ def combine(
...
         if True:  # not forward_mode.is_decode():
             if hidden_states.shape[0] > 0:
-                hidden_states = self.get_restored_hidden_states_by_experts(
-                    hidden_states
+                num_tokens = self.src2dst.shape[0] // self.router_topk
+                output = torch.zeros(
+                    (num_tokens, hidden_states.shape[1]),
+                    device=hidden_states.device,
+                    dtype=hidden_states.dtype,
+                )
```
Comment (Edenzzzz) [inline python/sglang/srt/layers/moe/ep_moe/token_dispatcher.py]: Use torch.empty?

File: python/sglang/srt/models/deepseek_v2.py [hunk 8]
```diff
@@ -294,7 +294,7 @@ def forward_deepep(
                 correction_bias=self.correction_bias,
             )
         if self.tp_size > 1:
-            recv_hidden_states, topk_idx, topk_weights, tokens_per_expert = (
+            recv_hidden_states, reorder_topk_ids, seg_indptr = (
```
Comment (Edenzzzz) [inline python/sglang/srt/models/deepseek_v2.py:297]: Should we add some short comments on the meaning/examples of reorder_topk_ids and seg_indptr for readability?

**Comment by Huixxi** [comment]
Will there be further optimization plans for this permute kernel？

**Comment by xutizhou** [comment]
> Will there be further optimization plans for this permute kernel？

We will continue to optimize the permute kernel, but it is not our top priority at the moment.

**Comment by Huixxi** [comment]
> node 0
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code \
  --tp 16 --dp 16  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 0 \
  --enable-dp-attention --enable-deepep-moe \
  --disable-cuda-graph
node 1
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code \
  --tp 16 --dp 16  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 1 \
  --enable-dp-attention --enable-deepep-moe \
  --disable-cuda-graph


But, it seems that I can't reproduce the performance of deepseek on 2 * H800 x 8 with roce rdma. I don't know why.
image

**Comment by xutizhou** [comment]
> > node 0
> > python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code 
> >   --tp 16 --dp 16  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 0 
> >   --enable-dp-attention --enable-deepep-moe 
> >   --disable-cuda-graph
> > node 1
> > python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3 --trust-remote-code 
> >   --tp 16 --dp 16  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 1 
> >   --enable-dp-attention --enable-deepep-moe 
> >   --disable-cuda-graph
> 
> But, it seems that I can't reproduce the performance of deepseek on 2 * H800 x 8 with roce rdma. I don't know why. image

The observed issue could potentially be attributed to ROCE network configuration. To verify this hypothesis, we recommend running the inter-node communication test from DeepEP's validation suite, specifically the [internode connectivity check](https://github.com/deepseek-ai/DeepEP/blob/main/tests/test_internode.py)
